playbooks/operationalizing/how-full-is-my-cluster.adoc

```diff
@@ -0,0 +1,130 @@
+The knapsack problem is a https://en.wikipedia.org/wiki/NP-completeness[np-complete]https://en.wikipedia.org/wiki/NP-completeness[https://en.wikipedia.org/wiki/NP-completeness[ ]]problem, which means that the algorithm that solves it scales exponentially with n (pods) and m (nodes). For this reason when n and m are big enough, a human cannot do a better job than a machine at solving this problem.
+
+*Best Practice:* refrain from pinning pods to nodes using node selectors, because this interferes with the ability of OpenShift to optimize the allocation of pods and increase density. 
+
+=== Limits
+
+Limits** **determine the maximum amount of resources (CPU and/or memory) that can be used by a container at runtime. Setting a limit corresponds to passing to the docker run command the https://docs.docker.com/engine/reference/run/#/user-memory-constraints[*--memory]* parameter for memory limits and the https://docs.docker.com/engine/reference/run/#/cpu-quota-constraint[*--cpu-quota]* parameter for CPU limits. +
+This influences the https://en.wikipedia.org/wiki/Cgroups[cgroup] that is created around the given container, limiting the resources it can use. +
+For memory, docker assumes that if you ask for a given amount of RAM, you should also get twice as much swap size. This implies that if you have swap active, your container will actually see thrice as much memory as you have assigned to it, and it could potentially swap. 
+
+*Best Practice*: https://docs.openshift.com/container-platform/3.5/admin_guide/overcommit.html#disabling-swap-memory[disable swap for nodes] (note that beginning with OpenShift 3.6, the node service will not start if swap is enabled).
+
+Again if you don’t specify limits for your containers, then OpenShift assumes that they are unbounded and that the container can consume all the resources available on the node. +
+To get an idea of the actual resources available in a node you can run (this command requires https://docs.openshift.com/container-platform/latest/install_config/cluster_metrics.html[cluster metrics] to be installed) +
+oc adm top node --heapster-namespace=openshift-infra --heapster-scheme=https <node-name> +
+Which will output something as follows: +
+NAME             CPU(cores)   CPU%      MEMORY(bytes)   MEMORY%   
+192.168.99.100   1007m        50%       5896Mi          49%
+
+This is the same node of the above example. Notice that the actual consumption of resources is much higher than what OpenShift estimated. The difference is due to the fact that in this node there are several pods that do not declare requests and limits.
+
+In summary, in order to fully being able to describe the available capacity of a cluster we should be able to answer two questions (and we need the answer them for at least memory and CPU):
+ . How much capacity OpenShift estimates is available based on the declared requests by the pods.
+ . How much capacity is really available based on current usage.
```

Comment:
To make these an ordered list, add a blank space between the paragraph and the list items, like the following:
```
In summary, in order to fully being able to describe the available capacity of a cluster we should be able to answer two questions (and we need the answer them for at least memory and CPU):

 . How much capacity OpenShift estimates is available based on the declared requests by the pods?
 . How much capacity is really available based on current usage?
```
Also questions should have questions marks.

Reply:
done